flink-libraries/flink-table/src/test/scala/org/apache/flink/table/ExpressionReductionTest.scala

```diff
@@ -155,15 +155,15 @@ class ExpressionReductionTest extends TableTestBase {
       "DataSetCalc",
       batchTableNode(0),
       term("select",
-        "13 AS _c0",
+        "CAST(13) AS _c0",
```

Comment:
Just played around a little bit. I think the problem is that the advanced types are not properly canonized. Using the following diff can pass all tests in `ArrayTypeTest`:

```
--- a/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala
+++ b/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala
@@ -133,12 +133,18 @@ class FlinkTypeFactory(typeSystem: RelDataTypeSystem) extends JavaTypeFactoryImp
   override def createTypeWithNullability(
     relDataType: RelDataType,
     nullable: Boolean)
-  : RelDataType = relDataType match {
-    case composite: CompositeRelDataType =>
-      // at the moment we do not care about nullability
-      composite
-    case _ =>
-      super.createTypeWithNullability(relDataType, nullable)
+  : RelDataType = {
+    val t = relDataType match {
+      case composite: CompositeRelDataType =>
+        // at the moment we do not care about nullability
+        composite
+      case array: ArrayRelDataType =>
+        val elementType = canonize(createTypeWithNullability(array.getComponentType, nullable))
+        new ArrayRelDataType(array.typeInfo, elementType, nullable)
+      case _ =>
+        super.createTypeWithNullability(relDataType, nullable)
+    }
+    canonize(t)
   }
 }
```

GroupWindowTest is still failing as it misses an identity projection. I'm wondering why `ProjectRemoveRule.INSTANCE` did not kick in...